### DISCUSSION TITLE
Look up item in cache

### DISCUSSION THREAD

**Discussion by sarink** [urql-graphql/urql#2454] [Q&A] [answered]
How do I look up an item in the cache? eg, `cache.find({ __typename: 'Todo', id: 1 })`? I know I could use `readFragment` and write a query, but I'm interested in obtaining _all_ the fields the cache has. Is it possible to just find an entity without constructing a graphql query or explicitly picking specific fields one at a time (a la `cache.resolve`)?

**Comment by JoviDeCroock** [comment, marked as answer]
You can use `cache.inspectFields` which will give you all fields for an entity. so `cache.inspectQuery('Query')` can give you all the queries you have done and `cache.inspectFields({ __typename: 'Todo', id: 1 })` can be used to get all keys for that todo. This can be used to automate `cache.resolve`. Generally we advise on using `readFragment` and `readQuery`.

**Comment by sarink** [comment]
@JoviDeCroock, thanks, I've decided to use a fragment instead, but I can't seem to get it to work. Do you know what could be going wrong here? 
```
          const id = '5mlJVdQ3kJf1G66y3D86Cqo7xKwRrgSGD3JWQ';

          const fromFragment = cache.readFragment(
            gql`
              fragment _ on Transaction {
                id
                name
              }
            `,
            'Transaction',
            { id }
          );

          const fromResolve = cache.resolve({ __typename: 'Transaction', id }, 'name');

          console.log('fragment', fromFragment); // null
          console.log('resolved', fromResolve); // 'correct name'
```

**Reply by JoviDeCroock** [reply]
Yes use 

```
          const fromFragment = cache.readFragment(
            gql`
              fragment _ on Transaction {
                id
                name
              }
            `,
            { id }
          );
```

**Comment by sarink** [comment]
That works 🎊  thanks

What tripped me up was when applying the generic,

```
gql`
  fragment TransactionParts on Transaction {
    id
    name
  }
`;

const fromFragment = cache.readFragment<TransactionPartsFragment>(
  TransactionPartsFragmentDoc,
  { id }, // Error, name is not specified
);
```

TS now thinks that all the entity fields specified on the fragment are required (which caused me to double check, then misunderstand, the method signature 😅 ). What's the recommended way to make it typesafe? The above types were generated using graphql-code-generator with the typescript-urql plugin
